cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,456 @@
+			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotFoundResponse(), errors.New("got an error")
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{*newFakeFirewall(fakeWorkerFWName, fakeInboundRule)}, newFakeOKResponse(), nil
+			},
+			expectedFirewall: newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+		},
+		{
+			name:    "get firewall from API with cached firewall ID",
+			fwCache: newFakeFirewallCache(fakeInboundRule),
+			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(fakeWorkerFWName, fakeInboundRule), newFakeOKResponse(), nil
+			},
+			expectedFirewall: newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+		},
+		{
+			name:    "when cache does not exist",
+			fwCache: newFakeFirewallCache(fakeInboundRule),
+			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeOKResponse(), nil
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return nil, newFakeOKResponse(), nil
+			},
+		},
+	}
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					getFunc:  test.expectedGodoFirewallGetResp,
+					listFunc: test.expectedGodoFirewallListResp,
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+
+			fw, err := fwManagerOp.Get(ctx)
+			if want, got := test.expectedError, err; want != got {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+
+			if diff := cmp.Diff(test.expectedFirewall, fw); diff != "" {
+				t.Errorf("Get() mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestFirewallController_Set(t *testing.T) {
+	testcases := []struct {
+		name                             string
+		fwCache                          firewallCache
+		inboundRules                     []godo.InboundRule
+		expectedGodoFirewallAddRulesResp func(context.Context, string, *godo.FirewallRulesRequest) (*godo.Response, error)
+		expectedGodoFirewallCreateResp   func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallGetResp      func(context.Context, string) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallListResp     func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallUpdateResp   func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error)
+		expectedError                    error
+		expectedFirewall                 *godo.Firewall
+	}{
+		{
+			name:    "create firewall when cache does not exist (i.e. initial startup)",
```

Comment:
```
fwCache: newFakeFirewallCacheEmpty(),
firewallRequest: &godo.FirewallRequest{
	Name:          testWorkerFWName,
	InboundRules:  testInboundRules,
	OutboundRules: testOutboundRules,
	Tags:          testWorkerFWTags,
},
```

The inbound rules are populated now. I think this might be an outdated comment.

Reply:
Yep, it's about `firewallRequest` now which you added, thanks. 👍